machineconfiguration/v1/types_machineosbuild.go

```diff
@@ -0,0 +1,208 @@
+	// This field describes the location of the final image, which will be pushed by the build once complete.
+	// The format of the image push spec is: host[:port][/namespace]/name:<tag> or svc_name.namespace.svc[:port]/repository/name:<tag>.
+	// The length of the push spec must be between 1 to 447 characters.
+	// +required
+	RenderedImagePushSpec ImageTagFormat `json:"renderedImagePushSpec"`
+}
+
+// MachineOSBuildStatus describes the state of a build and other helpful information.
+// +kubebuilder:validation:XValidation:rule="has(self.buildEnd) ? has(self.buildStart) && timestamp(self.buildStart) < timestamp(self.buildEnd) : true",message="buildEnd must be after buildStart"
+type MachineOSBuildStatus struct {
+	// conditions are state related conditions for the build. Valid types are:
+	// Prepared, Building, Failed, Interrupted, and Succeeded.
+	// Once a Build is marked as Failed or Interrupted, no future conditions can be set.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MaxItems=8
+	// +kubebuilder:validation:XValidation:rule="oldSelf.exists(x, x.type=='Failed') ? self==oldSelf : true",message="once a Failed condition is set, conditions are immutable"
+	// +kubebuilder:validation:XValidation:rule="oldSelf.exists(x, x.type=='Interrupted') ? self==oldSelf : true",message="once an Interrupted condition is set, conditions are immutable"
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// builder describes the image builder backend used for this build.
+	// +optional
+	Builder *MachineOSBuilderReference `json:"builder"`
+	// relatedObjects is a list of references to ephemeral objects such as ConfigMaps or Secrets that are meant to be consumed while the build process runs.
+	// After a successful build or when this MachineOSBuild is deleted, these ephemeral objects should be deleted.
+	// However, in the event of a failed build, the objects will not be deleted to allow for inspection and debugging of the failed build process.
+	// +kubebuilder:validation:MaxItems=10
+	// +listType=map
+	// +listMapKey=name
+	// +listMapKey=resource
+	// +optional
+	RelatedObjects []ObjectReference `json:"relatedObjects,omitempty"`
+	// buildStart is the timestamp corresponding to the build controller initiating the build backend for this MachineOSBuild.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="buildStart is immutable once set"
+	// +optional
+	BuildStart *metav1.Time `json:"buildStart"`
+	// buildEnd is the timestamp corresponding to completion of the builder backend.
+	// When omitted the build has either not been started, or is in progress.
+	// It will be populated once the build completes, fails or is interrupted.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="buildEnd is immutable once set"
```

Comment:
To honour the description, don't you need immutability to be relaxed here to allow transitioning from unset? e.g. oldSelf == null || self == oldSelf

Reply:
This works fine generally for all cases. I don't think the validation kicks in until `self` is not null. I've added a test case just in case but it does work.

Reply:
Yep, Jerry is correct, since this is a transition rule (contains oldSelf), it isn't effected until there is an `oldSelf`.

In the future, we may use `optionalOldSelf` to create ratcheting rules, that would change things, but for now, safe assumption